## 2d. Evidence - Resource Utilization QAS Measurements

Now we collect stored, CPU and memory usage data when predicting with the model, for the Performance scenario.

In [ ]:
{
    "tags": [],
    "quality_attribute": "",
    "description": "",
    "inputs": "",
    "output": ""
}

### Initialize MLTE Context

MLTE contains a global context that manages the currently active _session_. Initializing the context tells MLTE how to store all of the artifacts that it produces. This import will also set up global constants related to folders and model to use.

In [1]:
# Sets up context for the model being used, sets up constants related to folders and model data to be used.
from session import *

Creating initial custom lists at URI: local:///Users/rbrowersinning/Documents/ResearchFolders/Continuum_LTP/GitRepos/mlte_llm/demo/GardenBuddy/../store
Loaded 7 qa_categories for initial list
Loaded 30 quality_attributes for initial list
Creating sample catalog at URI: StoreType.LOCAL_FILESYSTEM:local:///Users/rbrowersinning/Documents/ResearchFolders/Continuum_LTP/GitRepos/mlte_llm/demo/GardenBuddy/../store
Loading sample catalog entries.
Loaded 9 entries for sample catalog.


### Set up scenario test case

In [ ]:
from mlte.negotiation.artifact import NegotiationCard

card = NegotiationCard.load()
qa = 3
print(card.quality_scenarios[qa].identifier)
print(card.quality_scenarios[qa].quality)
print(
    card.quality_scenarios[qa].stimulus,
    "from ",
    card.quality_scenarios[qa].source,
    " during ",
    card.quality_scenarios[qa].environment,
    ". ",
    card.quality_scenarios[qa].response,
    card.quality_scenarios[qa].measure,
)

card.default-qas_004
Resource  Utilization
The model running on the loaned device receives pictures taken at the garden from  the Garden Buddy application  during  normal operations .  The loaned devices are small, inexpensive devices with limited CPU power, as well as limited memory and disk space (512 MB and 150 MB, respectively). The model executes on the loaned device without any errors due to unavailable resources


**A Specific test case generated from the scenario:**

**Data and Data Source:**	The original test dataset can be used (i.e., original test data set aside for testing during data split into training, validation, and test).

**Measurement and Condition:**	    
    1- Executing the model on the loaned platform will not exceed maximum CPU usage of 30% to ensure reasonable response time. CPU usage will be measure using ps.
    2- Memory usage at inference time will not exceed available memory of 512 MB. This will be measured using pmap.
    3- Disk usage will not exceed allocated disk space of 150 MB bytes. This will be measured by adding the size of each file in the path for the model code.

**Context:**	Normal Operation

### Measurements

Prepare and execute measurements.

In [4]:
from mlte.measurement.storage import LocalObjectSize
from mlte.evidence.types.real import Real
from mlte.measurement.units import Units

store_measurement = LocalObjectSize("model size")
size: Real = store_measurement.evaluate(MODELS_DIR, unit=Units.byte)
print(size)
size.save(force=True)

206705557.0 byte


ArtifactModel(header=ArtifactHeaderModel(identifier='evidence.model size', type='evidence', timestamp=1762179519, creator=None, level='version'), body=EvidenceModel(artifact_type=<ArtifactType.EVIDENCE: 'evidence'>, metadata=EvidenceMetadata(test_case_id='model size', measurement=MeasurementMetadata(measurement_class='mlte.measurement.storage.local_object_size.LocalObjectSize', output_class='mlte.evidence.types.real.Real', additional_data={})), evidence_class='mlte.evidence.types.real.Real', value=RealValueModel(evidence_type=<EvidenceType.REAL: 'real'>, real=206705557.0, unit='byte')))

In [5]:
from mlte.measurement.process_measurement import ProcessMeasurement
from mlte.measurement.cpu import LocalProcessCPUUtilization, CPUStatistics

cpu_measurement = LocalProcessCPUUtilization("predicting cpu")
cpu_stats: CPUStatistics = cpu_measurement.evaluate(MODEL_COMMAND)
print(cpu_stats)
cpu_stats.save(force=True)

Deleting log
TensorFlow version: 2.18.0
Found 1 files.
Size of model file (/Users/rbrowersinning/Documents/ResearchFolders/Continuum_LTP/GitRepos/mlte_llm/demo/GardenBuddy/model/oxford_flower_model.keras): 206705544 bytes
Memory used for the entire model loading process: 555089920 bytes.
Running inference on 1 samples...
(1, 224, 224, 3)
Model - Input Validation Pass - RGB image loaded
1/1 ━━━━━━━━━━━━━━━━━━━━ 1s 518ms/step
Model - Output Validation Pass - (1, 102) - validated
Model - file: /Users/rbrowersinning/Documents/ResearchFolders/Continuum_LTP/GitRepos/mlte_llm/demo/GardenBuddy/data/sample predicted to be lotus (class [55]) with confidence [0.02620832]

--- STATISTICS ---
Average elapsed time per inference: 0.56560 seconds
Average memory used per inference: 41467904.00000 bytes.
Average: 0.5796666666666667 percent
Minimum: 0.0 percent
Maximum: 4.8469999999999995 percent


ArtifactModel(header=ArtifactHeaderModel(identifier='evidence.predicting cpu', type='evidence', timestamp=1762179543, creator=None, level='version'), body=EvidenceModel(artifact_type=<ArtifactType.EVIDENCE: 'evidence'>, metadata=EvidenceMetadata(test_case_id='predicting cpu', measurement=MeasurementMetadata(measurement_class='mlte.measurement.cpu.local_process_cpu_utilization.LocalProcessCPUUtilization', output_class='mlte.measurement.cpu.local_process_cpu_utilization.CPUStatistics', additional_data={})), evidence_class='mlte.measurement.cpu.local_process_cpu_utilization.CPUStatistics', value=OpaqueValueModel(evidence_type=<EvidenceType.OPAQUE: 'opaque'>, data={'avg': 0.5796666666666667, 'min': 0.0, 'max': 4.8469999999999995, 'unit': 'percent'})))

In [6]:
from mlte.measurement.memory import (
    LocalProcessMemoryUtilization,
    MemoryStatistics,
)

mem_measurement = LocalProcessMemoryUtilization("predicting memory")
mem_stats: MemoryStatistics = mem_measurement.evaluate(MODEL_COMMAND)
print(mem_stats)
mem_stats.save(force=True)

Deleting log
TensorFlow version: 2.18.0
Found 1 files.
Size of model file (/Users/rbrowersinning/Documents/ResearchFolders/Continuum_LTP/GitRepos/mlte_llm/demo/GardenBuddy/model/oxford_flower_model.keras): 206705544 bytes
Memory used for the entire model loading process: 565035008 bytes.
Running inference on 1 samples...
(1, 224, 224, 3)
Model - Input Validation Pass - RGB image loaded
1/1 ━━━━━━━━━━━━━━━━━━━━ 1s 533ms/step
Model - Output Validation Pass - (1, 102) - validated
Model - file: /Users/rbrowersinning/Documents/ResearchFolders/Continuum_LTP/GitRepos/mlte_llm/demo/GardenBuddy/data/sample predicted to be lotus (class [55]) with confidence [0.02620832]

--- STATISTICS ---
Average elapsed time per inference: 0.54956 seconds
Average memory used per inference: 36847616.00000 bytes.
Average: 455528 kilobyte
Minimum: 32 kilobyte
Maximum: 1080704 kilobyte


ArtifactModel(header=ArtifactHeaderModel(identifier='evidence.predicting memory', type='evidence', timestamp=1762179555, creator=None, level='version'), body=EvidenceModel(artifact_type=<ArtifactType.EVIDENCE: 'evidence'>, metadata=EvidenceMetadata(test_case_id='predicting memory', measurement=MeasurementMetadata(measurement_class='mlte.measurement.memory.local_process_memory_utilization.LocalProcessMemoryUtilization', output_class='mlte.measurement.memory.local_process_memory_utilization.MemoryStatistics', additional_data={})), evidence_class='mlte.measurement.memory.local_process_memory_utilization.MemoryStatistics', value=OpaqueValueModel(evidence_type=<EvidenceType.OPAQUE: 'opaque'>, data={'avg': 455528, 'min': 32, 'max': 1080704, 'unit': 'kilobyte'})))

We can also avoid starting the training process twice by using the asynch methods for both measurements. We start the training process once and pass the id to both measurements.

In [7]:
from mlte.measurement.cpu import LocalProcessCPUUtilization
from mlte.measurement.memory import LocalProcessMemoryUtilization
from mlte.measurement.process_measurement_group import ProcessMeasurementGroup

# Create measurement group
measurements = ProcessMeasurementGroup()

# Add measurements to group.
measurements.add(LocalProcessCPUUtilization("predicting cpu"))
measurements.add(LocalProcessMemoryUtilization("predicting memory"))

# Evaluate the measurements.
evidences = measurements.evaluate(command=MODEL_COMMAND)

# Get results.
cpu_stats = evidences["predicting cpu"]
mem_stats = evidences["predicting memory"]

# Inspect values
print(cpu_stats)
print(mem_stats)

# Save to artifact store
cpu_stats.save(force=True)
mem_stats.save(force=True)

Deleting log
TensorFlow version: 2.18.0
Found 1 files.
Size of model file (/Users/rbrowersinning/Documents/ResearchFolders/Continuum_LTP/GitRepos/mlte_llm/demo/GardenBuddy/model/oxford_flower_model.keras): 206705544 bytes
Memory used for the entire model loading process: 541442048 bytes.
Running inference on 1 samples...
(1, 224, 224, 3)
Model - Input Validation Pass - RGB image loaded
1/1 ━━━━━━━━━━━━━━━━━━━━ 1s 527ms/step
Model - Output Validation Pass - (1, 102) - validated
Model - file: /Users/rbrowersinning/Documents/ResearchFolders/Continuum_LTP/GitRepos/mlte_llm/demo/GardenBuddy/data/sample predicted to be lotus (class [55]) with confidence [0.02620832]

--- STATISTICS ---
Average elapsed time per inference: 0.54392 seconds
Average memory used per inference: 38780928.00000 bytes.
Average: 1.6158333333333335 percent
Minimum: 0.0 percent
Maximum: 5.032 percent
Average: 462920 kilobyte
Minimum: 736 kilobyte
Maximum: 1059504 kilobyte


ArtifactModel(header=ArtifactHeaderModel(identifier='evidence.predicting memory', type='evidence', timestamp=1762179565, creator=None, level='version'), body=EvidenceModel(artifact_type=<ArtifactType.EVIDENCE: 'evidence'>, metadata=EvidenceMetadata(test_case_id='predicting memory', measurement=MeasurementMetadata(measurement_class='mlte.measurement.memory.local_process_memory_utilization.LocalProcessMemoryUtilization', output_class='mlte.measurement.memory.local_process_memory_utilization.MemoryStatistics', additional_data={})), evidence_class='mlte.measurement.memory.local_process_memory_utilization.MemoryStatistics', value=OpaqueValueModel(evidence_type=<EvidenceType.OPAQUE: 'opaque'>, data={'avg': 462920, 'min': 736, 'max': 1059504, 'unit': 'kilobyte'})))